<launch>
  <!-- robot state publisher -->
  <arg name="urdf_file" />
  <!-- for base nodes -->
  <arg name="base_frame_id" default="base_link" />
  <arg name="odom_frame_id" default="odom" />
  <arg name="robot_yaml_filepath" />
  <arg name="odom_yaml_filepath" />
  <arg name="imu_adjust_param_filepath" />
  <!-- points down sampler -->
  <arg name="raw_points_topic" />
  <arg name="downsampler_node_name" />
  <!-- ndt matching -->
  <arg name="common_yaml_filepath" />
  <arg name="ndt_matching_yaml_filepath" />
  <arg name="mcl_3dl_yamlpath" />
  <arg name="start_course_idx" />
  <arg name="course_config_yamlpath" />
  <arg name="status_management_yamlpath" />
  <!-- bag replay -->
  <arg name="bagfile_path" />
  <arg name="start" />
  <arg name="duration" />
  <arg name="rate" />
  <arg name="wait" />
  <!-- Result dump file -->
  <arg name="dump_pose_file" default="False" />
  <arg name="localized_pose_filepath" default="" />
  <arg name="pose_pair_filepath" default="" />
  <arg name="kml_filepath" default="" />
  <!-- rviz -->
  <arg name="rviz_config_file" />

  <!-- Load common parameters -->
  <rosparam command="load" file="$(arg common_yaml_filepath)" />

  <!-- Base nodes bringup -->
  <include file="$(find scripts)/launch/system/basenodes_bringup.launch" >
    <arg name="urdf_file" value="$(arg urdf_file)" />
    <arg name="base_frame_id" value="$(arg base_frame_id)" />
    <arg name="odom_frame_id" value="$(arg odom_frame_id)" />
    <arg name="robot_yaml_filepath" value="$(arg robot_yaml_filepath)" />
    <arg name="odom_yaml_filepath" value="$(arg odom_yaml_filepath)" />
    <arg name="imu_adjust_param_filepath" value="$(arg imu_adjust_param_filepath)" />
    <arg name="rviz_config_file" value="$(arg rviz_config_file)" />
  </include>

  <include file="$(find scripts)/launch/system/localizer_bringup.launch">
    <arg name="raw_points_topic" value="$(arg raw_points_topic)" />
    <arg name="downsampler_node_name" value="$(arg downsampler_node_name)" />
    <arg name="ndt_matching_yaml_filepath" value="$(arg ndt_matching_yaml_filepath)" />
    <arg name="mcl_3dl_yamlpath" value="$(arg mcl_3dl_yamlpath)" />
  </include>

  <!-- Status maanger bringup -->
  <include file="$(find scripts)/launch/modules/common/status_management.launch">
    <arg name="start_course_idx" value="$(arg start_course_idx)" />
    <arg name="course_config_yamlpath" value="$(arg course_config_yamlpath)" />
    <arg name="status_management_yamlpath" value="$(arg status_management_yamlpath)" />
  </include>

  <group if="$(eval dump_pose_file)" >
    <node pkg="scripts" type="pose_writer.py" name="pose_writer" output="log" required="True" 
          args="-localized_pose_filepath $(arg localized_pose_filepath) -pose_pair_filepath $(arg pose_pair_filepath) -kml_filepath $(arg kml_filepath)" />
  </group>

  <include file="$(find scripts)/launch/modules/common/rosbag_replay.launch">
    <arg name="bagfile_path" value="$(arg bagfile_path)" />
    <arg name="start" value="$(arg start)" />
    <arg name="duration" value="$(arg duration)" />
    <arg name="rate" value="$(arg rate)" />
    <arg name="wait" value="$(arg wait)" />
  </include>

</launch>